File: engine/src/modules/processors/control_value_visualization.cpp
/*
  Copyright (C) 2026 Joshua Wade

  This file is part of Anthem.

  Anthem is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Anthem is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Anthem. If not, see <https://www.gnu.org/licenses/>.
*/

#include "control_value_visualization.h"

#include "modules/core/engine.h"
#include "modules/core/visualization/visualization_broker.h"
#include "modules/processing_graph/runtime/node_process_context.h"

#include <algorithm>
#include <cmath>

namespace anthem {

std::optional<NumericVisualizationData> ControlValueVisualizationProvider::getTypedData() {
  return latestValue.drainLatest();
}

void ControlValueVisualizationProvider::rt_pushValue(double value, int64_t sampleTimestamp) {
  latestValue.rt_set(value, sampleTimestamp);
}

ControlValueVisualizationProcessor::ControlValueVisualizationProcessor(
    const ControlValueVisualizationProcessorModelImpl& _impl)
  : Processor("ControlValueVisualization"), ControlValueVisualizationProcessorModelBase(_impl) {}

ControlValueVisualizationProcessor::~ControlValueVisualizationProcessor() {
  unregisterVisualizationProvider();
}

void ControlValueVisualizationProcessor::initialize(
    std::shared_ptr<ModelBase> selfModel, std::shared_ptr<ModelBase> parentModel) {
  ControlValueVisualizationProcessorModelBase::initialize(selfModel, parentModel);

  registerVisualizationProvider();
}

void ControlValueVisualizationProcessor::prepareToProcess(ProcessorPrepareCallback complete) {
  complete(std::nullopt);
}

std::optional<TimestampedVisualizationValue<double>>
ControlValueVisualizationProcessor::rt_getBlockValue(
    const juce::AudioSampleBuffer* inputBuffer, int numSamples, int64_t blockStartSample) {
  if (inputBuffer == nullptr || numSamples <= 0 || inputBuffer->getNumChannels() <= 0) {
    return std::nullopt;
  }

  const int availableSamples = std::min(numSamples, inputBuffer->getNumSamples());
  if (availableSamples <= 0) {
    return std::nullopt;
  }

  const int latestSampleIndex = availableSamples - 1;
  const double rawValue = static_cast<double>(inputBuffer->getReadPointer(0)[latestSampleIndex]);
  if (!std::isfinite(rawValue)) {
    return std::nullopt;
  }

  return TimestampedVisualizationValue<double>{
      .sampleTimestamp = blockStartSample + availableSamples,
      .value = std::clamp(rawValue, 0.0, 1.0),
  };
}

void ControlValueVisualizationProcessor::process(NodeProcessContext& context, int numSamples) {
  auto* inputBuffer = context.getInputControlBuffer(
      ControlValueVisualizationProcessorModelBase::controlInputPortId);
  const int64_t blockStartSample = Engine::getInstance().transport->rt_sampleCounter;
  auto value = rt_getBlockValue(inputBuffer, numSamples, blockStartSample);

  auto* provider = visualizationProvider.rt_getProvider();
  if (value.has_value() && provider != nullptr) {
    provider->rt_pushValue(value->value, value->sampleTimestamp);
  }
}

void ControlValueVisualizationProcessor::registerVisualizationProvider() {
  if (visualizationId().empty()) {
    return;
  }

  jassert(visualizationProvider.rt_getProvider() == nullptr);
  visualizationProvider =
      RegisteredVisualizationProvider<ControlValueVisualizationProvider>::registerDataProvider(
          visualizationId(), std::make_unique<ControlValueVisualizationProvider>());
}

void ControlValueVisualizationProcessor::unregisterVisualizationProvider() {
  visualizationProvider.release();
}

} // namespace anthem

File: lib/model/processing_graph/processors/control_value_visualization.dart
/*
  Copyright (C) 2026 Joshua Wade

  This file is part of Anthem.

  Anthem is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Anthem is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Anthem. If not, see <https://www.gnu.org/licenses/>.
*/

import 'package:anthem/helpers/id.dart';
import 'package:anthem/helpers/project_entity_id_allocator.dart';
import 'package:anthem/model/processing_graph/node.dart';
import 'package:anthem/model/processing_graph/node_port.dart';
import 'package:anthem/model/processing_graph/node_port_config.dart';
import 'package:anthem/model/processing_graph/processors/processor.dart';
import 'package:anthem/model/project_model_getter_mixin.dart';
import 'package:anthem_codegen/include.dart';
import 'package:mobx/mobx.dart';

part 'control_value_visualization.g.dart';

/// A control-value visualization sink.
///
/// This node reads a control input and publishes the latest normalized value as
/// a visualization stream. It is intended to tap the end of generated
/// automation chains without affecting the destination parameter.
@AnthemModel.syncedModel(
  cppBehaviorClassName: 'ControlValueVisualizationProcessor',
  cppBehaviorClassIncludePath:
      'modules/processors/control_value_visualization.h',
)
class ControlValueVisualizationProcessorModel
    extends _ControlValueVisualizationProcessorModel
    with
        Processor,
        _$ControlValueVisualizationProcessorModel,
        _$ControlValueVisualizationProcessorModelAnthemModelMixin {
  ControlValueVisualizationProcessorModel({
    required super.nodeId,
    required super.visualizationId,
  });

  ControlValueVisualizationProcessorModel.create({
    required ProjectEntityIdAllocator idAllocator,
    required super.visualizationId,
  }) : super(nodeId: idAllocator.allocateId());

  ControlValueVisualizationProcessorModel.uninitialized()
    : super(nodeId: -1, visualizationId: '');

  factory ControlValueVisualizationProcessorModel.fromJson(
    Map<String, dynamic> json,
  ) => _$ControlValueVisualizationProcessorModelAnthemModelMixin.fromJson(json);

  @override
  NodeModel createNode() {
    return NodeModel(
      id: nodeId,
      processor: this,
      controlInputPorts: AnthemObservableList.of([
        NodePortModel(
          nodeId: nodeId,
          id: controlInputPortId,
          config: NodePortConfigModel(dataType: NodePortDataType.control),
        ),
      ]),
    );
  }

  static int get controlInputPortId =>
      _ControlValueVisualizationProcessorModel.controlInputPortId;

  static String buildVisualizationId({
    required Id nodeId,
    required int portId,
  }) {
    return 'automation-control-value-$nodeId-$portId';
  }
}

abstract class _ControlValueVisualizationProcessorModel
    with Store, AnthemModelBase, ProjectModelGetterMixin {
  static const int controlInputPortId = 0;

  Id nodeId;

  /// Visualization stream ID used for the published control values.
  @anthemObservable
  String visualizationId;

  _ControlValueVisualizationProcessorModel({
    required this.nodeId,
    required this.visualizationId,
  });
}
